p2p/ping.go

```diff
@@ -0,0 +1,160 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/libp2p/go-libp2p/p2p/protocol/ping"
+	ma "github.com/multiformats/go-multiaddr"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	pingInterval = time.Second
+)
+
+type pingStat struct {
+	numSuccess int
+	numFail    int
+}
+
+type Ping struct {
+	sync.Mutex
+	logger log.Log
+	h      host.Host
+	peers  []peer.ID
+	pr     routing.PeerRouting
+	stats  map[peer.ID]*pingStat
+	ctx    context.Context
+	cancel context.CancelFunc
+	eg     errgroup.Group
+}
+
+func NewPing(logger log.Log, h host.Host, peers []peer.ID, pr routing.PeerRouting) *Ping {
+	p := &Ping{
+		logger: logger,
+		h:      h,
+		peers:  peers,
+		pr:     pr,
+		stats:  make(map[peer.ID]*pingStat),
+	}
+	return p
+}
+
+func (p *Ping) Start() {
+	if p.cancel != nil {
+		return
+	}
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+	for _, peer := range p.peers {
+		p.startForPeer(peer)
+	}
+}
+
+func (p *Ping) doPing(ctx context.Context, peerID peer.ID) (<-chan ping.Result, error) {
+	_, err := p.pr.FindPeer(ctx, peerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find peer %s: %w", peerID, err)
+	}
+
+	return ping.Ping(ctx, p.h, peerID), nil
+}
+
+func (p *Ping) startForPeer(peerID peer.ID) {
+	p.h.ConnManager().Protect(peerID, "sm-ping")
+
+	p.eg.Go(func() error {
+		var ctx context.Context
+		var cancel context.CancelFunc
+		defer func() {
+			if cancel != nil {
+				cancel()
+			}
+		}()
+	OUTER:
+		for {
+			if cancel != nil {
+				cancel()
+			}
+			ctx, cancel = context.WithCancel(p.ctx)
+			ch, err := p.doPing(ctx, peerID)
+			if err != nil {
+				cancel()
+				p.logger.With().Error("pre-PING connect failed", log.Err(err))
+				select {
+				case <-p.ctx.Done():
+					return p.ctx.Err()
+				case <-time.After(pingInterval):
+					continue
+				}
+			}
```

Comment:
A `context.CancelFunc` can be called any amount of times (all but the first call are no-ops) and calling `cancel` of a child context isn't necessary when the parent context was already canceled. The code could be shortened to:

```suggestion
func (p *Ping) startForPeer(ctx context.Context, peerID peer.ID) {
	p.h.ConnManager().Protect(peerID, "sm-ping")

	p.eg.Go(func() error {
	OUTER:
		for {
			pingCtx, cancel := context.WithCancel(ctx)
			defer cancel() // this servers to avoid a dead go routine if we leave the scope and forget calling cancel explicitly
			ch, err := p.doPing(pingCtx, peerID)
			if err != nil {
				cancel()
				p.logger.With().Error("pre-PING connect failed", log.Err(err))
				select {
				case <-ctx.Done():
					return ctx.Err()
				case <-time.After(pingInterval):
					continue
				}
			}
```

Reply:
Some extra cancels were added due to linter complaints about missing cancels on some of the code paths

Reply:
Ping code cleaned up